sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
+   * pull the build iterator to build a bigger batch on the GPU. This is not optimized
+   * because we hold onto the semaphore during the entire time after realizing the goal
+   * has been hit.
+   *
+   * @param hostTargetBatchSize target batch size goal on the host
+   * @param buildOutput output attributes of the build plan
+   * @param buildIter build iterator
+   * @param streamIter stream iterator
+   * @param spillCallback metric updater in case downstream iterators spill
+   * @param coalesceMetricsMap metrics map with metrics to be used in downstream
+   *                           iterators
+   * @return a pair of `ColumnarBatch` and streamed iterator that can be
+   *         used for the join
+   */
+  def getBuiltBatchAndStreamIter(
+      hostTargetBatchSize: Long,
+      buildOutput: Seq[Attribute],
+      buildIter: Iterator[ColumnarBatch],
+      streamIter: Iterator[ColumnarBatch],
+      spillCallback: SpillCallback,
+      coalesceMetricsMap: Map[String, GpuMetric]): (ColumnarBatch, Iterator[ColumnarBatch]) = {
+    val semWait = coalesceMetricsMap(GpuMetric.SEMAPHORE_WAIT_TIME)
+    val buildTime = coalesceMetricsMap(GpuMetric.BUILD_TIME)
+    var bufferedBuildIterator: CloseableBufferedIterator[ColumnarBatch] = null
+    closeOnExcept(bufferedBuildIterator) { _ =>
+      val startTime = System.nanoTime()
+      // find if the build side is non-empty, and if the first batch is
+      // a serialized batch. If neither condition is met, we fallback to the
+      // `getSingleBatchWithVerification` method.
+      val firstBatchIsSerialized = {
+        if (!buildIter.hasNext) {
+          false
+        } else {
+          bufferedBuildIterator = new CloseableBufferedIterator(buildIter.buffered)
+          val firstBatch = bufferedBuildIterator.head
+          if (firstBatch.numCols() != 1) {
+            false
+          } else {
+            firstBatch.column(0) match {
+              case _: SerializedTableColumn => true
+              case _ => false
+            }
+          }
+        }
+      }
+
+      if (!firstBatchIsSerialized) {
+        // fallback if we failed to find serialized build batches
+        val builtBatch =
+          ConcatAndConsumeAll.getSingleBatchWithVerification(
+            Option(bufferedBuildIterator).getOrElse(buildIter), buildOutput)
+        val delta = System.nanoTime() - startTime
+        buildTime += delta
+        (builtBatch, streamIter)
+      } else {
+        val dataTypes = buildOutput.map(_.dataType).toArray
+        closeOnExcept(new CloseableBufferedIterator(streamIter.buffered)) { bufferedStreamIter =>
+          val hostConcatIter = new MaybeHostConcatResultIterator(bufferedBuildIterator,
+              hostTargetBatchSize, dataTypes, coalesceMetricsMap)
+          closeOnExcept(hostConcatIter.next()) { maybeHostConcatResult =>
+            val buildBatch = if (!hostConcatIter.hasNext()) {
+              val singleBatchBufferTimeDelta = System.nanoTime() - startTime
```

Comment:
It's difficult to follow code when there's a `val somevar = if (...) {` followed by screens of code for the `if` body.  Suggest factoring this into a private method that is responsible for getting the build batch, whether that's using the single host buffer or falling back.

Reply:
Ok this should be better here: https://github.com/NVIDIA/spark-rapids/pull/4588/commits/a607e73887ffd05ad863d1af4a464b2a3999ea4a.